ophyd/controls/device.py

```diff
@@ -0,0 +1,439 @@
+                 add_prefix=None, doc=None, **kwargs):
+        self.attr = None  # attr is set later by the device when known
+        self.cls = cls
+        self.kwargs = kwargs
+        self.lazy = lazy
+        self.suffix = suffix
+        self.doc = doc
+        self.trigger_value = trigger_value  # TODO discuss
+
+        if add_prefix is None:
+            add_prefix = ('suffix', 'write_pv')
+
+        self.add_prefix = tuple(add_prefix)
+
+    def get_pv_name(self, instance, attr, suffix):
+        '''Get pv name for a given suffix'''
+        if attr in self.add_prefix:
+            # Optionally use a separator from the instance
+            if hasattr(instance, '_sep'):
+                sep = instance._sep
+            else:
+                sep = ''
+
+            return sep.join((instance.prefix, suffix))
+        else:
+            return suffix
+
+    def create_component(self, instance):
+        '''Create a component for the instance'''
+        kwargs = self.kwargs.copy()
+        kwargs['name'] = '{}.{}'.format(instance.name, self.attr)
+
+        for kw in self.add_prefix:
+            # If any keyword arguments need a prefix, tack it on
+            if kw in kwargs:
+                suffix = self.get_pv_name(instance, kw, kwargs[kw])
+                kwargs[kw] = suffix
+
+        # Otherwise, we only have suffix to update
+        pv_name = self.get_pv_name(instance, 'suffix', self.suffix)
+
+        cpt_inst = self.cls(pv_name, parent=instance, **kwargs)
+
+        if self.lazy and hasattr(self.cls, 'wait_for_connection'):
+            cpt_inst.wait_for_connection()
+
+        return cpt_inst
+
+    def make_docstring(self, parent_class):
+        if self.doc is not None:
+            return self.doc
+
+        return '{} component with suffix {}'.format(self.attr, self.suffix)
+
+    def __get__(self, instance, owner):
+        if instance is None:
+            return self
+
+        if self.attr not in instance._signals:
+            instance._signals[self.attr] = self.create_component(instance)
+
+        return instance._signals[self.attr]
+
+    def __set__(self, instance, owner):
+        raise RuntimeError('Use .put()')
+
+
+class DynamicDeviceComponent:
+    '''An OphydDevice component that is a dynamically created sub-device'''
+
+    def __init__(self, defn, clsname=None, doc=None):
+        self.defn = defn
+        self.clsname = clsname
+        self.attr = None  # attr is set later by the device when known
+        self.lazy = False
+        self.doc = doc
+
+        # TODO: component compatibility
+        self.trigger_value = None
+        self.attrs = list(defn.keys())
+
+    def make_docstring(self, parent_class):
+        if self.doc is not None:
+            return self.doc
+
+        return '{} dynamicdevice containing {}'.format(self.attr,
+                                                       self.attrs)
+
+    def get_separator(self, instance):
```

Comment:
I think I left this as-is since I wasn't sure if the concept of the separator should exist at all. It makes the PV generation from prefix + sep + suffix slightly more confusing.